### PR TITLE
XGBoost - Add evals argument to fit to support eval on datasets other than training

### LINKED ISSUE #526: [FEA] Add ability to run evaluation on other `Dataset`(s) while training
# 🚀 Feature request

It would be great to be able to evaluate the results of the `xgboost` model in training on other datasets than train. This addresses the inherent ability of `xgboost` to overfit and would align with how `xgboost` is nearly always used (with early stopping).

Functionality:
- the ability to pass in a `Dataset` to be used for evaluation and to use that `Dataset` for early stopping
- the ability to specify whether to perform the evaluation on the passed in `Dataset`(s) only or whether to include running the metrics on train as well (the reason we might want to separate these two scenarios is that train sets can be very large and running some metrics can be quite expensive (which would lead to increased train time). Also, from a UX perspective, it is very easy to run out of horizontal space when training `xgboost` models and having more than one line of results per iteration doesn't make for a very nice experience (things become hard to follow and read).

### PR DESCRIPTION
**Pull request by oliverholworthy**
Fixes #526 

### Goals :soccer:

A common pattern when using XGBoost is to pass a different dataset (e.g. validation data) to be monitored when training. We would like to support this functionality by allowing users of the XGBoost class to specify these datasets

### Implementation Details :construction:

- Adding new parameter evals to the fit method which accepts a list of tuples similar to the evals expected by the xgboost train function. The difference is that we expcect a Merlin Dataset instead of a DMatrix here.

### Testing Details :mag:

Test added for custom eval. We check the evals result returned by the training.

### CODE REVIEW

**Review by radekosmulski — CHANGES_REQUESTED** [review]
Our implementation dies with a hard to understand error when the eval dataset name is not a single string.

Other than that looks good!

image

**Review by radekosmulski — CHANGES_REQUESTED** [review]

File: merlin/models/xgb/__init__.py [hunk 1]
```diff
@@ -123,14 +126,33 @@ def fit(
...
+        train_res = xgb.dask.train(
             self.dask_client, self.params, dtrain, evals=watchlist, **train_kwargs
-        )["booster"]
-        self.booster = booster
+        )
+        self.booster: xgb.Booster = train_res["booster"]
+        self.evals_result = train_res["history"]
 
-        return booster
+        return self.booster
```
[the remark this answers is not in the document]
Comment (oliverholworthy) [reply merlin/models/xgb/__init__.py:157]: I was unsure what to return here too. The Merlin Models tensorflow API and Keras return a keras history callback from the fit method which allows you to inspect the metrics that were computed while running.

I suppose the closest equivalent here would be the evals_result dictionary. Or we need to think about what would be consistent across the whole project.

**Review by radekosmulski — APPROVED** [review]
